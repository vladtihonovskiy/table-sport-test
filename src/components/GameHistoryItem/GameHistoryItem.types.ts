export default interface IGameHistoryItemProps {
  id: string;
  parentName: string;
  guestName: string;
  parentPoints: number;
  guestPoints: number;
  isPlayed: boolean;
}
