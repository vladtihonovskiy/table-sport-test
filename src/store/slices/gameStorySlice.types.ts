export interface playersObjectTypes {
  name: string;
  played: number;
  win: number;
  draw: number;
  lost: number;
  totalPoints: number;
}

export interface GameObjectType {
  id: string;
  parentName: string;
  guestName: string;
  parentPoints: number;
  guestPoints: number;
  isPlayed: boolean;
}

export interface GameStoryState {
  listOPlayers: playersObjectTypes[];
  listOfGames: GameObjectType[];
}

export interface SaveGameScorePayload {
  id: string;
  parentPoints: number;
  guestPoints: number;
  parentName: string;
  guestName: string;
}
