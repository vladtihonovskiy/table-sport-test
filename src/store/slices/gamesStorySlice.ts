import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import {
  GameObjectType,
  GameStoryState,
  playersObjectTypes,
  SaveGameScorePayload,
} from "./gameStorySlice.types";

const newPlayerObjectCreator = (name: string) => ({
  name,
  played: 0,
  win: 0,
  draw: 0,
  lost: 0,
  totalPoints: 0,
});

const newGameObjectCreator = (parentName: string, guestName: string) => ({
  id: `${parentName} ${guestName}`,
  parentName,
  guestName,
  guestPoints: 0,
  parentPoints: 0,
  isPlayed: false,
});

const createNewGames = (
  listOfPlayers: playersObjectTypes[],
  newCommandName: string
) => {
  const newGames: GameObjectType[] = [];
  listOfPlayers.forEach((item) => {
    newGames.push(newGameObjectCreator(item.name, newCommandName));
  });
  return newGames;
};

const initialState: GameStoryState = {
  listOPlayers: [],
  listOfGames: [],
};

export const gameStorySlice = createSlice({
  name: "gameStory",
  initialState,
  reducers: {
    addNewPlayerName: (state, action: PayloadAction<string>) => {
      if (state.listOPlayers.length >= 1) {
        const newListOfGames = createNewGames(
          [...state.listOPlayers],
          action.payload
        );
        state.listOfGames.push(...newListOfGames);
      }
      state.listOPlayers.push(newPlayerObjectCreator(action.payload));
    },
    saveGameResult: (state, action: PayloadAction<SaveGameScorePayload>) => {
      const { id, parentPoints, guestPoints, parentName, guestName } =
        action.payload;
      const updateGameIndex = state.listOfGames.findIndex(
        (item) => item.id === id
      );
      state.listOfGames[updateGameIndex] = {
        ...state.listOfGames[updateGameIndex],
        parentPoints,
        guestPoints,
        isPlayed: true,
      };

      // const parentPlayerIndex = state.listOPlayers.findIndex(
      //   (item) => item.name === parentName
      // );
      // const guestPlayerIndex = state.listOPlayers.findIndex(
      //   (item) => item.name === guestName
      // );
    },
  },
});

// Action creators are generated for each case reducer function
export const { addNewPlayerName, saveGameResult } = gameStorySlice.actions;

export default gameStorySlice.reducer;
