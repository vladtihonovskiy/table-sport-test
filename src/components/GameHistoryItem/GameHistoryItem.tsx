import React, { useCallback, useMemo, useState } from "react";
import { Box, Button, Grid, TextField, Typography } from "@mui/material";
import IGameHistoryItemProps from "./GameHistoryItem.types";
import { useAppDispatch } from "../../hooks/useAppDispatch";
import { saveGameResult } from "../../store/slices/gamesStorySlice";

export const GameHistoryItem: React.FC<IGameHistoryItemProps> = ({
  id,
  guestPoints,
  parentPoints,
  parentName,
  guestName,
  isPlayed,
}: IGameHistoryItemProps) => {
  const dispatch = useAppDispatch();

  const [parentPointsInput, setParentPointsInput] = useState(
    isPlayed ? parentPoints : "0"
  );
  const [guestPointsInput, setGuestPointsInput] = useState(
    isPlayed ? guestPoints : "0"
  );

  const onSaveHandleClick = useCallback(() => {
    dispatch(
      saveGameResult({
        id,
        parentPoints: +parentPointsInput,
        guestPoints: +guestPointsInput,
        parentName,
        guestName,
      })
    );
  }, [id, parentPointsInput, guestPointsInput, parentName, guestName]);

  const renderInputs = useMemo(
    () => (
      <>
        <TextField
          sx={{ maxWidth: 70, mx: 1 }}
          size="small"
          type="number"
          disabled={isPlayed}
          inputProps={{ min: 0, max: 10 }}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
            e.target.value = Math.max(0, parseInt(e.target.value, 10))
              .toString()
              .slice(0, 2);
            setParentPointsInput(+e.target.value);
          }}
          onKeyPress={(event) => {
            if (event?.key === "-" || event?.key === "+") {
              event.preventDefault();
            }
          }}
          value={+parentPointsInput}
        />
        <p>:</p>
        <TextField
          sx={{ maxWidth: 70, mx: 1 }}
          size="small"
          type="number"
          disabled={isPlayed}
          inputProps={{ min: 0, max: 10 }}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
            e.target.value = Math.max(0, parseInt(e.target.value, 10))
              .toString()
              .slice(0, 2);

            setGuestPointsInput(+e.target.value);
          }}
          value={+guestPointsInput}
          onKeyPress={(event) => {
            if (event?.key === "-" || event?.key === "+") {
              event.preventDefault();
            }
          }}
        />
      </>
    ),
    [isPlayed, parentPointsInput, guestPointsInput]
  );

  return (
    <Grid
      data-testid={`${id}-game-history-item`}
      container
      alignItems="center"
      sx={{
        flexDirection: { sm: "row" },
        height: { xs: "auto", sm: 50 },
        marginLeft: { xs: 0, sm: 10 },
        marginTop: { xs: 1, sm: 0 },
        justifyContent: { xs: "center", sm: "flex-start" },
      }}
    >
      <Box sx={{ m: { xs: 0, sm: 2 }, maxWidth: 60, width: "100%" }}>
        <Typography
          sx={{
            overflow: "hidden",
            textOverflow: "ellipsis",
            textAlign: "center",
          }}
        >
          {parentName}
        </Typography>
      </Box>
      {renderInputs}
      <Box sx={{ m: { xs: 0, sm: 2 }, maxWidth: 60, width: "100%" }}>
        <Typography
          sx={{
            overflow: "hidden",
            textOverflow: "ellipsis",
            textAlign: "center",
          }}
        >
          {guestName}
        </Typography>
      </Box>

      {/** show only if we didn't save the result * */}
      {!isPlayed && (
        <Button
          sx={{ width: { xs: "100%", sm: "auto" }, my: { xs: 1, sm: 0 } }}
          onClick={onSaveHandleClick}
          variant="contained"
        >
          Save
        </Button>
      )}
    </Grid>
  );
};

export default GameHistoryItem;
