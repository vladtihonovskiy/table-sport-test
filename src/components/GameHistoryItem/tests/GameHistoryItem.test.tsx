import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import * as reduxHooks from "react-redux";
import GameHistoryItem from "../GameHistoryItem";
import * as actions from "../../../store/slices/gamesStorySlice";

jest.mock("react-redux");

const mockedDispatch = jest.spyOn(reduxHooks, "useDispatch");

const defaultHistoryItem = {
  id: "Ukraine-Poland",
  parentName: "Ukraine",
  guestName: "Poland",
  guestPoints: 0,
  parentPoints: 0,
  isPlayed: false,
};

describe("GameHistoryItem", () => {
  it("should render", () => {
    const component = render(<GameHistoryItem {...defaultHistoryItem} />);

    expect(component).toMatchSnapshot();
  });

  it("should show correct names", () => {
    render(<GameHistoryItem {...defaultHistoryItem} />);
    expect(screen.getByText(defaultHistoryItem.parentName)).toBeInTheDocument();
    expect(screen.getByText(defaultHistoryItem.guestName)).toBeInTheDocument();
  });

  it("should button be in the component if game not played yet", () => {
    render(<GameHistoryItem {...defaultHistoryItem} />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  it("should call saveGameResult action on button click", () => {
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedSaveGameResult = jest.spyOn(actions, "saveGameResult");

    render(<GameHistoryItem {...defaultHistoryItem} />);
    const button = screen.getByRole("button");

    fireEvent.click(button);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(mockedSaveGameResult).toHaveBeenCalledWith({
      id: "Ukraine-Poland",
      parentName: "Ukraine",
      guestName: "Poland",
      guestPoints: 0,
      parentPoints: 0,
    });
  });
});
