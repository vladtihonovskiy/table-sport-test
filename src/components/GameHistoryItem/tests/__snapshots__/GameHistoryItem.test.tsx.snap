// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameHistoryItem should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container css-1dwu8wc-MuiGrid-root"
        data-testid="Ukraine-Poland-game-history-item"
      >
        <div
          class="MuiBox-root css-1b8v88t"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-uspqav-MuiTypography-root"
          >
            Ukraine
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1pa1qsc-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              max="10"
              min="0"
              type="number"
              value="0"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <p>
          :
        </p>
        <div
          class="MuiFormControl-root MuiTextField-root css-1pa1qsc-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              max="10"
              min="0"
              type="number"
              value="0"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiBox-root css-1b8v88t"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-uspqav-MuiTypography-root"
          >
            Poland
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-2gqfrj-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Save
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container css-1dwu8wc-MuiGrid-root"
      data-testid="Ukraine-Poland-game-history-item"
    >
      <div
        class="MuiBox-root css-1b8v88t"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-uspqav-MuiTypography-root"
        >
          Ukraine
        </p>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1pa1qsc-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            max="10"
            min="0"
            type="number"
            value="0"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <p>
        :
      </p>
      <div
        class="MuiFormControl-root MuiTextField-root css-1pa1qsc-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            max="10"
            min="0"
            type="number"
            value="0"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiBox-root css-1b8v88t"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-uspqav-MuiTypography-root"
        >
          Poland
        </p>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-2gqfrj-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Save
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
